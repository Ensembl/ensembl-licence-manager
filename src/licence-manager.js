const addLicence = (fileContent, licenceText, config) => {
  if (!shouldAddLicence(fileContent, licenceText)) {
    return fileContent;
  }
  const licenceComment = prepareLicenceComment(licenceText, config);
  return `${licenceComment}\n\n${fileContent}`;
};

const removeLicence = (fileContent, licenceText) => {
  const licenceComment = prepareLicenceComment(licenceText);
  const fileWithoutLicence = fileContent.replace(licenceComment, '');
  return fileWithoutLicence.replace(/^\s+/, ''); // remove leading white space
};

const replaceLicence = (fileContent, oldLicence, newLicence) => {
  const cleanFileContent = removeLicence(fileContent, oldLicence);
  return addLicence(cleanFileContent, newLicence);
};

const prepareLicenceComment = (licenceText, config) => {
  licenceText = licenceText.trim();
  const {
    prepend,
    append,
    eachLine: {
      prepend: prependEachLine = '',
      append: appendEachLine = ''
    }
  } = config;
  const lines = licenceText.split('\n');
  const result = [];

  if (prepend) {
    result.push(prepend);
  }

  for (line of lines) {
    line = `${prependEachLine}${line}${appendEachLine}`;
    result.push(line);
  }

  if (append) {
    result.push(append);
  }

  return result.join('\n');
};

const shouldAddLicence = (fileContent, licenceText) => {
  return !fileContent.includes(licenceText.trim());
};

module.exports = {
  addLicence,
  removeLicence,
  replaceLicence
};
