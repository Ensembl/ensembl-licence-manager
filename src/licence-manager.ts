import { Config } from './config';

const addLicence = (fileContent: string, licenceText: string, config: Config) => {
  if (!shouldAddLicence(fileContent, licenceText)) {
    return fileContent;
  }
  const licenceComment = prepareLicenceComment(licenceText, config);
  return `${licenceComment}\n\n${fileContent}`;
};

const removeLicence = (fileContent: string, licenceText: string, config: Config) => {
  const licenceComment = prepareLicenceComment(licenceText, config);
  const fileWithoutLicence = fileContent.replace(licenceComment, '');
  return fileWithoutLicence.replace(/^\s+/, ''); // remove leading white space
};

const replaceLicence = (fileContent: string, oldLicence: string, newLicence: string, config: Config) => {
  const cleanFileContent = removeLicence(fileContent, oldLicence, config);
  return addLicence(cleanFileContent, newLicence, config);
};

const prepareLicenceComment = (licenceText: string, config: Config) => {
  licenceText = licenceText.trim();
  const {
    prepend,
    append,
    eachLine: {
      prepend: prependEachLine = '',
      append: appendEachLine = ''
    }
  } = config;
  const lines = licenceText.split('\n');
  const result = [];

  if (prepend) {
    result.push(prepend);
  }

  for (let line of lines) {
    line = `${prependEachLine}${line}${appendEachLine}`;
    result.push(line);
  }

  if (append) {
    result.push(append);
  }

  return result.join('\n');
};

const shouldAddLicence = (fileContent: string, licenceText: string) => {
  return !fileContent.includes(licenceText.trim());
};

export {
  addLicence,
  removeLicence,
  replaceLicence
};
