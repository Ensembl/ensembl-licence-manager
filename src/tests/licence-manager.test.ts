import fs from 'fs';
import path from 'path';

import { addLicence, removeLicence, replaceLicence } from '../licence-manager';
import configs from '../config';

const licenceText = fs
  .readFileSync(path.resolve(__dirname, 'fixtures/lorem-ipsum.txt'), 'utf-8')
  .trim();
const newLicenceText = fs
  .readFileSync(path.resolve(__dirname, 'fixtures/zombie-ipsum.txt'), 'utf-8')
  .trim();
const sampleSourceFile = fs
  .readFileSync(path.resolve(__dirname, 'fixtures/example.js'), 'utf-8')
  .trim();

describe('licence-manager', () => {

  const config = configs.js;

  describe('addLicence', () => {
    it('adds licence to the head of the file', () => {
      expect(addLicence(sampleSourceFile, licenceText, config)).toMatchSnapshot();
    });

    it('does not modify file already containing licence', () => {
      const fileWithLicence = addLicence(sampleSourceFile, licenceText, config);
      expect(addLicence(fileWithLicence, licenceText, config)).toMatchSnapshot();
    });
  });

  describe('removeLicence', () => {
    it('removes licence from the head of the file', () => {
      const fileWithLicence = addLicence(sampleSourceFile, licenceText, config);
      expect(removeLicence(fileWithLicence, licenceText, config)).toBe(sampleSourceFile);
    });
  });

  describe('replaceLicence', () => {
    it('replaces old licence with the new one', () => {
      const fileWithLicence = addLicence(sampleSourceFile, licenceText, config);
      expect(replaceLicence(fileWithLicence, licenceText, newLicenceText, config)).toMatchSnapshot();
    });
  });


});
