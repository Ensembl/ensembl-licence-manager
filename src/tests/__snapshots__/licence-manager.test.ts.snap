// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`licence-manager addLicence adds licence to the head of the file 1`] = `
"/**
* Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et turpis
* ornare, feugiat ex non, viverra ipsum. Donec ac faucibus tortor, sed euismod
* nulla. Etiam eget dolor dui. Nullam feugiat vel massa sed commodo. Vivamus sed
* fermentum elit, mollis lobortis velit. Nam dignissim suscipit diam, vel
* facilisis elit. Sed ante mauris, sollicitudin eu finibus vel, consectetur vitae
* lectus. Sed a aliquet odio. Proin consectetur lacus elit, non posuere eros
* lobortis accumsan. Phasellus malesuada id velit eget aliquam. Aliquam sed lacus
* lacinia, accumsan eros eu, eleifend erat.
*
* Aenean molestie nibh a consequat tempus. Donec nec sollicitudin felis. Fusce
* felis sem, pharetra varius vestibulum et, fringilla at lectus. Vivamus id felis
* auctor, consequat purus varius, finibus nunc. Suspendisse faucibus placerat
* nunc nec lobortis. Mauris non enim dapibus, sagittis ipsum consectetur,
* placerat felis. Cras quis nulla varius, bibendum justo a, auctor nisi. Aliquam
* egestas, ligula quis volutpat scelerisque, lectus lectus ornare odio, eu
* condimentum mi ex sed purus. Curabitur dapibus vehicula turpis a vehicula.
* Praesent consectetur, nisi a imperdiet sodales, nisi nulla interdum nibh, nec
* tincidunt libero dolor non mauris. Fusce pellentesque diam vel tortor
* tincidunt, placerat sagittis nunc feugiat. Duis pharetra viverra erat, a
* suscipit odio. Vivamus eu consequat metus. Vestibulum a nibh elementum,
* sollicitudin nunc ut, faucibus enim. Pellentesque gravida massa est, ut tempor
* felis porttitor iaculis.
*
* Quisque dui ante, tincidunt et nunc ut, dignissim pharetra orci. Nam eget
* congue erat, eget auctor sem. Morbi ut tincidunt nibh. Nunc eget mollis erat.
* Mauris pulvinar semper feugiat. Praesent odio augue, tincidunt in tincidunt ut,
* condimentum a nunc. Ut blandit in augue et scelerisque.
*/

const foo = 1;
const bar = 2;

const add = (x, y) => x + y;

add(foo, bar);"
`;

exports[`licence-manager addLicence does not modify file already containing licence 1`] = `
"/**
* Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et turpis
* ornare, feugiat ex non, viverra ipsum. Donec ac faucibus tortor, sed euismod
* nulla. Etiam eget dolor dui. Nullam feugiat vel massa sed commodo. Vivamus sed
* fermentum elit, mollis lobortis velit. Nam dignissim suscipit diam, vel
* facilisis elit. Sed ante mauris, sollicitudin eu finibus vel, consectetur vitae
* lectus. Sed a aliquet odio. Proin consectetur lacus elit, non posuere eros
* lobortis accumsan. Phasellus malesuada id velit eget aliquam. Aliquam sed lacus
* lacinia, accumsan eros eu, eleifend erat.
*
* Aenean molestie nibh a consequat tempus. Donec nec sollicitudin felis. Fusce
* felis sem, pharetra varius vestibulum et, fringilla at lectus. Vivamus id felis
* auctor, consequat purus varius, finibus nunc. Suspendisse faucibus placerat
* nunc nec lobortis. Mauris non enim dapibus, sagittis ipsum consectetur,
* placerat felis. Cras quis nulla varius, bibendum justo a, auctor nisi. Aliquam
* egestas, ligula quis volutpat scelerisque, lectus lectus ornare odio, eu
* condimentum mi ex sed purus. Curabitur dapibus vehicula turpis a vehicula.
* Praesent consectetur, nisi a imperdiet sodales, nisi nulla interdum nibh, nec
* tincidunt libero dolor non mauris. Fusce pellentesque diam vel tortor
* tincidunt, placerat sagittis nunc feugiat. Duis pharetra viverra erat, a
* suscipit odio. Vivamus eu consequat metus. Vestibulum a nibh elementum,
* sollicitudin nunc ut, faucibus enim. Pellentesque gravida massa est, ut tempor
* felis porttitor iaculis.
*
* Quisque dui ante, tincidunt et nunc ut, dignissim pharetra orci. Nam eget
* congue erat, eget auctor sem. Morbi ut tincidunt nibh. Nunc eget mollis erat.
* Mauris pulvinar semper feugiat. Praesent odio augue, tincidunt in tincidunt ut,
* condimentum a nunc. Ut blandit in augue et scelerisque.
*/

const foo = 1;
const bar = 2;

const add = (x, y) => x + y;

add(foo, bar);"
`;

exports[`licence-manager replaceLicence replaces old licence with the new one 1`] = `
"/**
* Zombie ipsum reversus ab viral inferno, nam rick grimes malum cerebro. De carne
* lumbering animata corpora quaeritis. Summus brains sit​​, morbo vel maleficia?
* De apocalypsi gorger omero undead survivor dictum mauris. Hi mindless mortuis
* soulless creaturas, imo evil stalking monstra adventus resi dentevil vultus
* comedat cerebella viventium. Qui animated corpse, cricket bat max brucks
* terribilem incessu zomby. The voodoo sacerdos flesh eater, suscitat mortuos
* comedere carnem virus. Zonbi tattered for solum oculi eorum defunctis go lum
* cerebro. Nescio brains an Undead zombies. Sicut malus putrid voodoo horror.
* Nigh tofth eliv ingdead.
*
* Cum horribilem walking dead resurgere de crazed sepulcris creaturis, zombie
* sicut de grave feeding iride et serpens. Pestilentia, shaun ofthe dead scythe
* animated corpses ipsa screams. Pestilentia est plague haec decaying ambulabat
* mortuos. Sicut zeder apathetic malus voodoo. Aenean a dolor plan et terror
* soulless vulnerum contagium accedunt, mortui iam vivam unlife. Qui tardius
* moveri, brid eof reanimator sed in magna copia sint terribiles undeath
* legionis. Alii missing oculis aliorum sicut serpere crabs nostram. Putridi
* braindead odores kill and infect, aere implent left four dead.
*
* Lucio fulci tremor est dark vivos magna. Expansis creepy arm yof darkness ulnis
* witchcraft missing carnem armis Kirkman Moore and Adlard caeruleum in locis.
* Romero morbo Congress amarus in auras. Nihil horum sagittis tincidunt, zombie
* slack-jawed gelida survival portenta. The unleashed virus est, et iam zombie
* mortui ambulabunt super terram. Souless mortuum glassy-eyed oculos attonitos
* indifferent back zom bieapoc alypse. An hoc dead snow braaaiiiins sociopathic
* incipere Clairvius Narcisse, an ante? Is bello mundi z?
*/

const foo = 1;
const bar = 2;

const add = (x, y) => x + y;

add(foo, bar);"
`;
